File: packages/geospatial/distance.py
"""Geospatial distance and spatiotemporal proximity calculations.

Calculates physically accurate geodesic distances over the WGS84 sphere/ellipsoid
in SI meters, eliminating Euclidean degree distortion where 1 deg lat != 1 deg lon.
"""

import math
from datetime import datetime

from packages.geospatial.coordinates import validate_wgs84_coordinates

# IUGG recommended mean Earth radius in meters (R1)
WGS84_MEAN_EARTH_RADIUS_METERS: float = 6371008.8


def haversine_distance_meters(
    lat1: float,
    lon1: float,
    lat2: float,
    lon2: float,
) -> float:
    """Calculate the great-circle distance between two points on Earth in meters.

    Uses the Haversine formula with numerical clamping for precision and stability.

    Args:
        lat1: Latitude of point 1 in degrees [-90.0, 90.0].
        lon1: Longitude of point 1 in degrees [-180.0, 180.0].
        lat2: Latitude of point 2 in degrees [-90.0, 90.0].
        lon2: Longitude of point 2 in degrees [-180.0, 180.0].

    Returns:
        float: Geodesic distance in meters (non-negative).
    """
    lat1_v, lon1_v = validate_wgs84_coordinates(lat1, lon1)
    lat2_v, lon2_v = validate_wgs84_coordinates(lat2, lon2)

    # Identical coordinates short-circuit
    if lat1_v == lat2_v and lon1_v == lon2_v:
        return 0.0

    phi1 = math.radians(lat1_v)
    phi2 = math.radians(lat2_v)
    delta_phi = math.radians(lat2_v - lat1_v)
    delta_lambda = math.radians(lon2_v - lon1_v)

    sin_half_dphi = math.sin(delta_phi / 2.0)
    sin_half_dlambda = math.sin(delta_lambda / 2.0)

    a = (sin_half_dphi * sin_half_dphi) + (
        math.cos(phi1) * math.cos(phi2) * sin_half_dlambda * sin_half_dlambda
    )

    # Clamp to [0.0, 1.0] to guard against floating-point rounding errors near antipodes
    a_clamped = min(1.0, max(0.0, a))
    c = 2.0 * math.atan2(math.sqrt(a_clamped), math.sqrt(1.0 - a_clamped))

    return WGS84_MEAN_EARTH_RADIUS_METERS * c


def is_spatiotemporally_proximate(
    lat1: float,
    lon1: float,
    t1: datetime,
    lat2: float,
    lon2: float,
    t2: datetime,
    max_distance_meters: float,
    max_time_difference_hours: float,
) -> bool:
    """Evaluate whether two observations are proximate in both space and time.

    Args:
        lat1: Latitude of observation 1.
        lon1: Longitude of observation 1.
        t1: Timestamp of observation 1.
        lat2: Latitude of observation 2.
        lon2: Longitude of observation 2.
        t2: Timestamp of observation 2.
        max_distance_meters: Maximum spatial radius in meters.
        max_time_difference_hours: Maximum temporal delta in hours.

    Returns:
        bool: True if spatial distance <= radius AND time delta <= window.
    """
    if max_distance_meters < 0:
        raise ValueError(
            f"max_distance_meters must be non-negative, got {max_distance_meters}"
        )
    if max_time_difference_hours < 0:
        msg = (
            f"max_time_difference_hours must be non-negative, "
            f"got {max_time_difference_hours}"
        )
        raise ValueError(msg)

    # Check temporal threshold first (fast scalar comparison)
    time_diff_seconds = abs((t2 - t1).total_seconds())
    max_time_seconds = max_time_difference_hours * 3600.0
    if time_diff_seconds > max_time_seconds:
        return False

    # Check spatial threshold using geodesic distance
    distance_meters = haversine_distance_meters(lat1, lon1, lat2, lon2)
    return distance_meters <= max_distance_meters

File: apps/web/src/lib/geo/distance.ts
/**
 * Geodesic spatial calculation utilities based on the WGS-84 ellipsoid sphere approximation.
 */

const EARTH_RADIUS_METERS = 6371000; // Mean Earth radius in meters

/**
 * Calculates the great-circle geodesic distance between two points on Earth using the Haversine formula.
 * @param lat1 Latitude of point 1 in decimal degrees
 * @param lon1 Longitude of point 1 in decimal degrees
 * @param lat2 Latitude of point 2 in decimal degrees
 * @param lon2 Longitude of point 2 in decimal degrees
 * @returns Geodesic distance in meters
 */
export function calculateHaversineDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  if (
    typeof lat1 !== "number" ||
    typeof lon1 !== "number" ||
    typeof lat2 !== "number" ||
    typeof lon2 !== "number" ||
    isNaN(lat1) ||
    isNaN(lon1) ||
    isNaN(lat2) ||
    isNaN(lon2)
  ) {
    return 0;
  }

  const toRad = (deg: number) => (deg * Math.PI) / 180;

  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) *
      Math.cos(toRad(lat2)) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return Math.round(EARTH_RADIUS_METERS * c);
}

/**
 * Formats a metric distance into human-readable meters or kilometers.
 */
export function formatDistance(distanceMeters: number | null | undefined): string {
  if (distanceMeters === null || distanceMeters === undefined || isNaN(distanceMeters)) {
    return "N/A";
  }

  if (distanceMeters < 0) return "0 m";

  if (distanceMeters < 1000) {
    return `${Math.round(distanceMeters)} m`;
  }

  const km = distanceMeters / 1000;
  return `${km >= 10 ? km.toFixed(0) : km.toFixed(1)} km`;
}

/**
 * Evaluates operational industrial exposure tier based on geodesic proximity.
 */
export function calculateExposureLevel(
  distanceMeters: number | null | undefined
): "HIGH" | "MEDIUM" | "LOW" | "NONE" {
  if (distanceMeters === null || distanceMeters === undefined || isNaN(distanceMeters)) {
    return "NONE";
  }

  if (distanceMeters <= 500) {
    return "HIGH";
  }
  if (distanceMeters <= 2000) {
    return "MEDIUM";
  }
  return "LOW";
}
